File: proposed-projects/round-trip-experiment/reverse/20260909T001426Z_1c28aa33/20260910T041632Z_e61d8830/ported_implementation/gbnf/rules_builder/symbol_ids.py
from typing import Dict, Iterator, Optional, Tuple


class SymbolIds:
    def __init__(self) -> None:
        self._map: Dict[str, int] = {}
        self._pos: Dict[str, int] = {}
        self._reverse_map: Dict[int, str] = {}

    def entries(self) -> Iterator[Tuple[str, int]]:
        return iter(self._map.items())

    def __iter__(self) -> Iterator[Tuple[str, int]]:
        return iter(self._map.items())

    @property
    def size(self) -> int:
        return len(self._map)

    def get(self, key: str) -> Optional[int]:
        return self._map.get(key)

    def set(self, key: str, value: int, pos: int) -> None:
        self._map[key] = value
        self._reverse_map[value] = key
        self._pos[key] = pos

    def has(self, key: str) -> bool:
        return key in self._map

    def reverse_get(self, key: int) -> str:
        if key not in self._reverse_map:
            raise ValueError(f"SymbolIds does not contain value: {key}")
        return self._reverse_map[key]

    def get_pos(self, key: str) -> int:
        if key not in self._pos:
            raise ValueError(f"SymbolIds does not contain key: {key}")
        return self._pos[key]

File: packages/gbnf-experiment/data/20260909T182714Z_2643e4be/ported_implementation/src/rules-builder/symbol-ids.ts
import { KeyError, ValueError } from "../utils/errors/python-errors.js";

export class SymbolIds {
  #map: Map<string, number> = new Map();
  #pos: Map<string, number> = new Map();
  #reverseMap: Map<number, string> = new Map();

  items(): IterableIterator<[string, number]> {
    return this.#map.entries();
  }

  [Symbol.iterator](): IterableIterator<[string, number]> {
    return this.#map.entries();
  }

  get size(): number {
    return this.#map.size;
  }

  // Equivalent of Python's `symbol_ids[key]`, which raises on a missing key.
  get(key: string): number {
    const value = this.#map.get(key);
    if (value === undefined) {
      throw new KeyError(key);
    }
    return value;
  }

  set(key: string, value: number, pos: number): void {
    this.#map.set(key, value);
    this.#reverseMap.set(value, key);
    this.#pos.set(key, pos);
  }

  has(key: string): boolean {
    return this.#map.has(key);
  }

  reverseGet(key: number): string {
    const val = this.#reverseMap.get(key);
    if (val === undefined) {
      throw new ValueError(`SymbolIds does not contain value: ${key}`);
    }
    return val;
  }

  getPos(key: string): number {
    const val = this.#pos.get(key);
    if (val === undefined) {
      throw new ValueError(`SymbolIds does not contain key: ${key}`);
    }
    return val;
  }
}
